File: internal/model/db/projection/installment.go
package projection

import (
	"akatengu/internal/enums"

	"github.com/shopspring/decimal"
)

type Installment struct {
	InstallmentId   int64                   `db:"installment_id"`
	TransactionId   *int64                  `db:"txn_id"`
	LedgerId        int64                   `db:"ledger_id"`
	Description     string                  `db:"description"`
	TotalAmount     decimal.Decimal         `db:"total_amount"`
	TotalPeriods    int                     `db:"total_periods"`
	PaidPeriods     int                     `db:"paid_periods"`
	AmountPerPeriod decimal.Decimal         `db:"amount_per_period"`
	StartDate       string                  `db:"start_date"`
	EndDate         *string                 `db:"end_date"`
	InterestRate    decimal.Decimal         `db:"interest_rate"`
	InterestType    enums.InterestType      `db:"interest_type"`
	Status          enums.InstallmentStatus `db:"status"`
	Note            string                  `db:"note"`
}

type InstallmentPayment struct {
	PaymentId     int64                          `db:"payment_id"`
	InstallmentId int64                          `db:"installment_id"`
	TransactionId *int64                         `db:"txn_id"`
	Period        int                            `db:"period_no"`
	Amount        decimal.Decimal                `db:"amount"`
	Interest      decimal.Decimal                `db:"interest"`
	DueDate       string                         `db:"due_date"`
	PaidDate      *string                        `db:"paid_date"`
	Status        enums.InstallmentPaymentStatus `db:"status"`
}

File: frontside/src/types/installment.ts
export type InterestType             = 'FREE' | 'FIXED_RATE';
export type InstallmentStatus        = 'ACTIVE' | 'COMPLETED' | 'CANCELED';
export type InstallmentPaymentStatus = 'PENDING' | 'PAID';

export interface Installment {
  installment_id:    number;
  transaction_id:    number | null;
  ledger_id:         number;
  description:       string;
  total_amount:      string;
  total_periods:     number;
  paid_periods:      number;
  amount_per_period: string;
  start_date:        string;
  end_date:          string | null;
  interest_rate:     string;
  interest_type:     InterestType;
  status:            InstallmentStatus;
  note:              string;
  updated_by:        string | null;
  updated_at:        string | null;
}

export interface InstallmentPayment {
  payment_id:     number;
  installment_id: number;
  transaction_id: number | null;
  period:         number;
  amount:         string;
  interest:       string;
  due_date:       string;
  paid_date:      string | null;
  status:         InstallmentPaymentStatus;
  updated_by:     string | null;
  updated_at:     string | null;
}

export interface InstallmentCreatedPayload {
  amount:            string;
  installment_count: number;
  start_date:        string;
  interest_type:     InterestType;
  annual_rate:       string;
  account_id:        string;
  ledger_id:         number;
  memo:              string;
  note:              string;
}

export interface InstallmentPeriodPaidPayload {
  installment_id: number;
  period:         number;
  paid_date:      string;
  paid_ledger_id: number;
}